service/qos: Modularize service level controller to avoid invalid access to auth::service

dawmd opened the pull request:
Move management over effective service levels from `service_level_controller`
to a new dedicated type -- `auth_integration`.

### Problem

Before these changes, it was possible for the service level controller to try
to access `auth::service` after it was deinitialized. For instance, it could
happen when reloading the cache. That HAS happened as described in the following
issue: scylladb/scylladb#24792.

Although the problem might have been mitigated or even resolved in
scylladb/scylladb@10214e13bdf44f82c2d6a53cdcf4d0ae288c838a, it's not clear
how the service will be used in the future. It's best to prevent similar bugs
than trying to fix them later on.

The logic responsible for preventing to access an uninitialized `auth::service`
was also either non-existent, complex, or non-sufficient.

### Solution

To prevent accessing `auth::service` by the service level controller, we extract
the relevant portion of the code to a separate entity -- `auth_integration`.
It's an internal helper type whose sole purpose is to manage effective service
levels.

Thanks to that, we were able to nest the lifetime of `auth_integration` within
the lifetime of `auth::service`. It's now impossible to attempt to dereference
it while it's uninitialized.

If a bug related to an invalid access is spotted again, though, it might also
be easier to debug it now.

### Impact

There should be no visible change to the users of the interface of the service
level controller. We strived to make the patch minimal, and the only affected
part of the logic should be related to how `auth::service` is accessed.

The relevant portion of the initialization and deinitialization flow:

(a) Before the changes:

1. Initialize `service_level_controller`. Pass a reference to an uninitialized
   `auth::service` to it.
2. Initialize other services.
3. Initialize and start `auth::service`.
4. (work)
5. Stop and deinitialize `auth::service`.
6. Deinitialize other services.
7. Deinitialize `service_level_controller`.

(b) After the changes:

1. Initialize `service_level_controller`. Pass a reference to an uninitialized
   `auth::service` to it. (*)
2. Initialize other services.
3. Initialize and start `auth::service`.
4. Initialize `auth_integration`. Register it in `service_level_controller`.
5. (work)
6. Unregister `auth_integration` in `service_level_controller` and deinitialize
   it.
7. Stop and deinitialize `auth::service`.
8. Deinitialize other services.
9. Deinitialize `service_level_controller`.

(*):
    The reference to `auth::service` in `service_level_controller` is still
    necessary. We need to access the service when dropping a distributed
    service level.

    Although it would be best to cut that link between the service level
    controller and `auth::service` too, effectively separating the entities,
    it would require more work, so we leave it as-is for now.

    It shouldn't prove problematic as far as accessing an uninitialized service
    goes. Trying to drop a service level at the point when we're de-initializing
    auth should be impossible.

    For more context, see the function `drop_distributed_service_level` in
    `service_level_controller`.

### Testing

A trivial test has been included in the PR. Although its value is questionable
as we only try to reload the service level cache at a specific moment, it's
probably the best we can deliver to provide a reproducer of the issue this patch
is resolving.

Fixes scylladb/scylladb#24792

Backport: The impact of the bug was minimal as it only affected the shutdown.
However, since CI is failing because of it, let's backport the change to all
supported versions.

Jadw1 commented:
@andrzej-jackowski-scylladb Since recently you are also working with service levels, could you also review this PR?

dawmd commented on main.cc:
```diff
@@ -2276,6 +2276,13 @@ sharded<locator::shared_token_metadata> token_metadata;
 
             auth_service.start(std::move(perm_cache_config), std::ref(qp), std::ref(group0_client), std::ref(mm_notifier), std::ref(mm), auth_config, maintenance_socket_enabled::no, std::ref(hashing_worker)).get();
 
+            // Reproducer of scylladb/scylladb#24792.
```
That would unfortunately lead to really cursed code: `auth::service` doesn't have access to `service_level_controller` (as far as I can tell), so we won't be able to do that from there.
On the other hand, it's probably not possible from `service_level_controller` either without registering some kind of observer on `auth::service`.
Manipulating error injections at shutdown is also problematic: we could try to detect the moment when `auth::service` is stopped and try to trigger reloading the cache then, but the API we use to communicate with Scylla is already shut down by then.

dawmd commented:
> I'm concerned that no one will be able to guess what an effective_service_level_controller is and how it's different from service_level_controller.

I agree. For now, I left the name as-is in the cover letter. If we keep the solution as a separate service, I'll change its name in the next version of the PR.

> Can you explain the lifetimes here. Before the change, do the lifetimes not nest? If they do nest, which is the outer scope and which is the inner? And why do we have a use-after-deinit? And if they don't nest, can we nest them?
> [...]
> Maybe the solution you present solves this better, but I'd like an overview in the cover letter rather than wading into the patches.

I rewrote the cover letter. It should be clearer why the lifetimes don't nest and how the problem appears (although it's quite brief and uses a reference to an issue on GitHub). I can expand it further providing more details if needed.

> Perhaps this:
> 
> start service_level_controller
> start database
> start auth
> imbue auth into service_level_controller
> (work)
> drain auth from service_level_controller
> stop auth
> stop database
> stop service_level_controller

This patch does more or less that except that I extracted the relevant portion of the code from service level controller to a new service (submodule). Maybe it's an overkill, though, and it would suffice to simply register and unregister `auth::service` at the right moments: the premise would be similar to what it's like now (with registering and unregistering the effective service level). That would make the changes smaller and easier to review because there would be no moving code back and forth.

My original intention was to cut off the link between `service_level_controller` and `auth::service` completely, but as it turned out impossible without modifying a bit more code, that might be a more feasible approach.

avikivity commented:
Thanks, with the new cover letter it's abundantly clear.

I suggest to rename effective_service_level_controller to service_level_controll_auth_integration.

dawmd commented:
> Ok. Then maybe
> 
> basic_service_level_controller -> service_level_controller (like std::basic_string -> std::string).
> 
> Depends on what the majority of the users interact with.

I think I didn't explain it properly. We have two services, `service_level_controller` and `effective_service_level_controller`. The former has the latter as a component (that's registered as soon as `auth::service` is initialised, and unregistered just before `auth::service` is deinitialised), while the latter has a reference to the former to be able to operate. I guess it can be described as some kind of dynamic dependency injection.

Some ideas off the top of my head: `effective_service_level_accessor`, `service_level_auth_accessor`, `service_level_auth_manager`, `service_level_auth_controller`, `role_service_level_controller`, `role_service_level_accessor`, `role_service_level_manager`, `service_level_associator`.

avikivity commented:
> > Ok. Then maybe
> > basic_service_level_controller -> service_level_controller (like std::basic_string -> std::string).
> > Depends on what the majority of the users interact with.
> 
> I think I didn't explain it properly. We have two services, `service_level_controller` and `effective_service_level_controller`. The former has the latter as a component (that's registered as soon as `auth::service` is initialised, and unregistered just before `auth::service` is deinitialised), while the latter has a reference to the former to be able to operate. I guess it can be described as some kind of dynamic dependency injection.

This is the first option I mentioned:

> service level controller without the hooks into auth
additional service that provides the hooks into auth but doesn't intereact with the original service level controller users

But you said it was the other one. So now I'm more confused than before.


service_level_controller and effective_service_level_controller implies the second is a specialization of the first, not that thew second is a helper for the first that isn't visible to users (I prefer the helper approach, but even more I want to understand what's being proposed before wading into the patches).




> 
> Some ideas off the top of my head: `effective_service_level_accessor`, `service_level_auth_accessor`, `service_level_auth_manager`, `service_level_auth_controller`, `role_service_level_controller`, `role_service_level_accessor`, `role_service_level_manager`, `service_level_associator`, `service_level_role_associator`.

I proposed "service_level_controller_auth_integration" because it names the two services involved and the relationship between them (if it's indeed the not-user-visible helper approach).

piodul commented:
> It's the latter, though I personally think of it more as a submodule of service level controller.

The two services have different responsibilities.

- The currently-named effective_service_level_controller is concerned with answering the question: given a role, what are the parameters (timeout, scheduling group, etc.) that should be used? The actual parameters for the role (called in the code as "effective service levels") are computed in a relatively non-trivial way, the service levels of either role's ancestors or children all need to be taken into consideration in order to compute it.
- The service_level_controller handles other stuff: operations on service levels (adding, removing, listing), manages scheduling groups, etc.

Everything used to be bundled into the `service_level_controller`, but this PR splits out the part mentioned in the first point to a separate service.

If we moved management of scheduling groups into the `effective_service_level_controller` then I believe that those two services would have different concerns: the ESLC would be useful in the cql_server where the scheduling group needs to be assigned to the connection after login (and on service level update, i.e. `switch_tenant`), whereas stuff like adding/deleting/listing service levels is mostly needed by CQL statements that manipulate service levels - so more relevant to the `query_state` class.

piodul commented on service/qos/effective_service_level_controller.hh:
```diff
@@ -0,0 +1,102 @@
+};
+
+/// NOTICE: Dummy implementation for now -- work in progress.
+class effective_service_level_controller : public peering_sharded_service<effective_service_level_controller> {
```
> I also don't like having the cache in the name - correct me if I'm wrong, but I think the main purpose of the logic is to access the auth information.

@andrzej-jackowski-scylladb The goal of the "cache" is to accelerate the `find_effective_service_level()` queries. To calculate an effective service level for a given user you need to traverse its hierarchy (its ancestors or its children, not sure which) and take the parameters of all service levels in the hierarchy into account. We precompute this because it is cheaper to do it for all roles/SLs at once, the cache speeds up the lookup on connection establishment and it fits the "group0 reload from disk" model rather well (we go over all connections and update their parameters if there is a SL change).

I'm not sure if the "cache" name is the best, either but... I'm not sure that I have a better suggestion.

avikivity commented:
> > It's the latter, though I personally think of it more as a submodule of service level controller.
> 
> The two services have different responsibilities.
> 
> * The currently-named effective_service_level_controller is concerned with answering the question: given a role, what are the parameters (timeout, scheduling group, etc.) that should be used? The actual parameters for the role (called in the code as "effective service levels") are computed in a relatively non-trivial way, the service levels of either role's ancestors or children all need to be taken into consideration in order to compute it.

So effective_service_level_controller is (efffective_service_level)_controller, not effective_(service_level_controller)? May explain some of the confusion.

effective_service_level_controller acts as a plug-in database for service_level_controller? It's not a controller itself.


> * The service_level_controller handles other stuff: operations on service levels (adding, removing, listing), manages scheduling groups, etc.

The _controller part.

> 
> Everything used to be bundled into the `service_level_controller`, but this PR splits out the part mentioned in the first point to a separate service.
> 
> If we moved management of scheduling groups into the `effective_service_level_controller` then I believe that those two services would have different concerns: the ESLC would be useful in the cql_server where the scheduling group needs to be assigned to the connection after login (and on service level update, i.e. `switch_tenant`), whereas stuff like adding/deleting/listing service levels is mostly needed by CQL statements that manipulate service levels - so more relevant to the `query_state` class.

I don't think we should move anything. The "database plugin" concept is natural and makes sense. It's the name that suggests that something else is going on.

dawmd commented:
Version 2.0:
* Rethought the solution. I went with Avi's idea to only use the new service internally. Thanks to that, we keep the old interface as-is and the diff is much smaller. It should also be easier to follow the changes now.

dawmd commented:
@avikivity you convinced me that the approach with an internal helper (not facing the users) is a better idea, though I'm not sure if that wasn't your actual intention. In any case, I'm mentioning that because there are a few comments above that explain the design and thought behind the initial version of this PR, but it's no longer valid, so keep it in mind when reviewing the patch. I'll update the cover letter shortly.

dawmd commented:
Version 2.1:
* Got rid of sharding the new service. It's not necessary.

avikivity commented:
Looks good. My nit: the pull request title gives no clue about what's going on (solving non-nested lifetime problem)

dawmd commented on service/qos/service_level_controller.cc:
```diff
@@ -282,7 +282,11 @@ future<> service_level_controller::update_service_levels_cache(qos::query_contex
             }
-            _auth_integration->clear_cache();
+
+            if (_auth_integration) {
```
I wanted to keep the changes separate because they have different purposes. Introducing `_auth_integration` in `service_level_controller` moves code around (without changing the logic, at least not much), while adding the conditions (here and in the cache reloading code) alters the behaviour of the service. I think it would also be confusing to introduce one condition in one commit, while the other in another.

yaronkaikov commented:
This patch is breaking Gating in `aarch64` - https://jenkins.scylladb.com/job/scylla-master/job/next/9544/testReport/ , please dequeue it ASAP
@avikivity @denesb @scylladb/scylla-maint

piodul commented:
> This patch is breaking Gating in `aarch64` - https://jenkins.scylladb.com/job/scylla-master/job/next/9544/testReport/ , please dequeue it ASAP @avikivity @denesb @scylladb/scylla-maint

@yaronkaikov and @lsfreitas told me that they are bisecting right now (we are not 100% sure that this PR breaks next)

dawmd commented:
There is a bug.

```cpp
future<> service_level_controller::unregister_auth_integration() {
    SCYLLA_ASSERT(_auth_integration != nullptr);
    // First, prevent new tasks coming to `auth_integration`.
    auto tmp = std::exchange(_auth_integration, nullptr);
    // Now we can stop it.
    co_await tmp->stop();
}
```

We cannot set `_auth_integration` to `nullptr` until all ongoing tasks have finished. Otherwise, we can get invalid memory accesses. That's the only suspicious thing I see at the moment, but it shouldn't explain the failures on Jenkins (we run the tests in debug mode, so a sanitiser should scream).

dawmd commented:
> There is a bug.
> 
> ```c++
> future<> service_level_controller::unregister_auth_integration() {
>     SCYLLA_ASSERT(_auth_integration != nullptr);
>     // First, prevent new tasks coming to `auth_integration`.
>     auto tmp = std::exchange(_auth_integration, nullptr);
>     // Now we can stop it.
>     co_await tmp->stop();
> }
> ```
> 
> We cannot set `_auth_integration` to `nullptr` until all ongoing tasks have finished. Otherwise, we can get invalid memory accesses. That's the only suspicious thing I see at the moment, but it shouldn't explain the failures on Jenkins (we run the tests in debug mode, so a sanitiser should scream).

This was premature, and I think at least this piece of code is fine.
